connect/runtime/src/main/java/org/apache/kafka/connect/runtime/errors/LogReporter.java

```diff
@@ -0,0 +1,131 @@
+
+    private static final Logger log = LoggerFactory.getLogger(LogReporter.class);
+
+    public static final String PREFIX = "errors.log";
+
+    public static final String LOG_ENABLE = "enable";
+    public static final String LOG_ENABLE_DOC = "Log the error context along with the other application logs.";
+    public static final boolean LOG_ENABLE_DEFAULT = false;
+
+    public static final String LOG_INCLUDE_MESSAGES = "include.messages";
+    public static final String LOG_INCLUDE_MESSAGES_DOC = "Include the Connect Record which failed to process in the log.";
+    public static final boolean LOG_INCLUDE_MESSAGES_DEFAULT = false;
+
+    private final ConnectorTaskId id;
+
+    private LogReporterConfig config;
+    private ErrorHandlingMetrics errorHandlingMetrics;
+
+    public LogReporter(ConnectorTaskId id) {
+        this.id = id;
+    }
+
+    private static final ConfigDef CONFIG_DEF = new ConfigDef()
+                .define(LOG_ENABLE, ConfigDef.Type.BOOLEAN, LOG_ENABLE_DEFAULT, ConfigDef.Importance.MEDIUM, LOG_ENABLE_DOC)
+                .define(LOG_INCLUDE_MESSAGES, ConfigDef.Type.BOOLEAN, LOG_INCLUDE_MESSAGES_DEFAULT, ConfigDef.Importance.MEDIUM, LOG_INCLUDE_MESSAGES_DOC);
+
+    /**
+     * Log error context.
+     *
+     * @param context the processing context.
+     */
+    @Override
+    public void report(ProcessingContext context) {
+        if (!config.isEnabled()) {
+            return;
+        }
+
+        if (!context.failed()) {
+            return;
+        }
+
+        StringBuilder builder = message(context);
+        log.error(builder.toString(), context.error());
+        errorHandlingMetrics.recordErrorLogged();
+    }
+
+    @Override
+    public void setMetrics(ErrorHandlingMetrics errorHandlingMetrics) {
+        this.errorHandlingMetrics = errorHandlingMetrics;
+    }
+
+    // Visible for testing
+    StringBuilder message(ProcessingContext context) {
+        StringBuilder builder = new StringBuilder("Error encountered in task ");
+        builder.append(id).append(" while performing ");
+        builder.append(context.stage().name());
+        builder.append(" operation with class '");
+        builder.append(context.executingClass());
+        builder.append('\'');
+        if (config.canLogMessages() && context.sourceRecord() != null) {
+            builder.append(", with record = ");
+            builder.append(context.sourceRecord());
+        } else if (config.canLogMessages() && context.consumerRecord() != null) {
+            ConsumerRecord<byte[], byte[]> msg = context.consumerRecord();
+            builder.append(", msg.topic='").append(msg.topic()).append('\'');
+            builder.append(", msg.partition=").append(msg.partition());
+            builder.append(", msg.offset=").append(msg.offset());
+            builder.append(", msg.timestamp=").append(msg.timestamp());
+            builder.append(", msg.timestampType=").append(msg.timestampType());
```

Comment:
Do we need to prefix these with `msg.`?

Also, it might be nice to control the message format when message keys and values are to be included. I know that's more code, but it could be made to be similar to the other option.

Reply:
Is there any way to put this into methods on the `ProcessingContext`, like `toString(boolean includeRecord)`? That way, its `toString()` method could really just be `toString(false)`. Not only is it a bit more testable, the logic of this reporter is more readable and focused on writing to the log.